### ISSUE TITLE
Legion runtime assert fails with multiple CPU processors and no utility procs

### ISSUE THREAD

**Issue by streichler** [StanfordLegion/legion#236]
In both master and stable,
```
cd tutorial/01_tasks_and_futures
make LG_RT_DIR=../../runtime -j4
for i in {1..1000}; do echo $i; ./tasks_and_futures -ll:util 0 -ll:cpu 4 || break; done
```
fails pretty often (usually after only a few tries) with:
```
Computing the first 7 Fibonacci numbers...
Fibonacci(0) = 0 (elapsed = 0.00 s)
Fibonacci(1) = 1 (elapsed = 0.00 s)
Fibonacci(2) = 1 (elapsed = 0.01 s)
Fibonacci(3) = 2 (elapsed = 0.02 s)
Fibonacci(4) = 3 (elapsed = 0.02 s)
Fibonacci(5) = 5 (elapsed = 0.03 s)
tasks_and_futures: ../../runtime/legion/legion_tasks.cc:2271: void Legion::Internal::TaskOp::trigger_children_committed(): Assertion `children_complete' failed.
Aborted (core dumped)
```

Failures happen less often with fewer CPU procs (`-ll:util 0 -ll:cpu 1` fails after ~100 tries), and never occur if there is only a single CPU proc.  Failures also never seem to occur if utility procs exist.

**Comment by lightsighter** [comment]
Should be fixed with a9f00ee. Assigning back to @streichler to confirm.

**Comment by streichler** [comment]
The assert fail is gone, but in 1000 runs, I got a single failure that looks like the use of a corrupted handle in a call to `Reservation::acquire`, so I suspect there may still be some sort of race condition here.  I'm trying to reproduce it (and get a backtrace).  Leaving the bug assigned to me for now.

**Comment by streichler** [comment]
Here we go:
```
4207
Computing the first 7 Fibonacci numbers...
Fibonacci(0) = 0 (elapsed = 0.00 s)
Fibonacci(1) = 1 (elapsed = 0.00 s)
Fibonacci(2) = 1 (elapsed = 0.04 s)
Fibonacci(3) = 2 (elapsed = 0.09 s)
Fibonacci(4) = 3 (elapsed = 0.12 s)
Fibonacci(5) = 5 (elapsed = 0.16 s)
Fibonacci(6) = 8 (elapsed = 0.15 s)
tasks_and_futures: ../../runtime/realm/runtime_impl.cc:1852: Realm::ReservationImpl* Realm::RuntimeImpl::get_lock_impl(Realm::ID): Assertion `0' failed.
BACKTRACE (0, 7f3db2f60700)
----------
  ./tasks_and_futures : Realm::RuntimeImpl::realm_backtrace(int)+0x6b
  /lib/x86_64-linux-gnu/libc.so.6 : ()+0x36cb0
  /lib/x86_64-linux-gnu/libc.so.6 : gsignal()+0x37
  /lib/x86_64-linux-gnu/libc.so.6 : abort()+0x148
  /lib/x86_64-linux-gnu/libc.so.6 : ()+0x2fbf6
  /lib/x86_64-linux-gnu/libc.so.6 : ()+0x2fca2
  ./tasks_and_futures : Realm::RuntimeImpl::get_lock_impl(Realm::ID)+0x18c
  ./tasks_and_futures : Realm::Reservation::try_acquire(bool, unsigned int, bool, Realm::Event) const+0x5e
  ./tasks_and_futures : Legion::AutoLock::AutoLock(Realm::Reservation, unsigned int, bool, Legion::RtEvent)+0x46
  ./tasks_and_futures : Legion::Internal::Runtime::unregister_distributed_collectable(unsigned long long)+0x5f
  ./tasks_and_futures : Legion::Internal::DistributedCollectable::unregister_with_runtime(Legion::Internal::VirtualChannelKind) const+0x7c
  ./tasks_and_futures : Legion::Internal::FutureImpl::~FutureImpl()+0x5c
  ./tasks_and_futures : void Legion::Internal::legion_delete<Legion::Internal::FutureImpl>(Legion::Internal::FutureImpl*)+0x1f
  ./tasks_and_futures : Legion::Future::operator=(Legion::Future const&)+0x4d
  ./tasks_and_futures : Legion::Internal::IndividualTask::deactivate()+0x219
  ./tasks_and_futures : Legion::Internal::Operation::commit_operation(bool, Legion::RtEvent)+0x26a
  ./tasks_and_futures : Legion::Internal::IndividualTask::trigger_task_commit()+0x146
  ./tasks_and_futures : Legion::Internal::TaskOp::trigger_children_committed()+0xef
  ./tasks_and_futures : Legion::Internal::InnerContext::post_end_task(void const*, unsigned long, bool)+0x4e6
  ./tasks_and_futures : Legion::Internal::InnerContext::end_task(void const*, unsigned long, bool)+0x724
  ./tasks_and_futures : Legion::LegionSerialization::end_helper(Legion::Runtime*, Legion::Internal::ContextInterface*, void const*, unsigned long, bool)+0x3e
  ./tasks_and_futures : Legion::LegionSerialization::StructHandler<int, false>::end_task(Legion::Runtime*, Legion::Internal::ContextInterface*, int*)+0x33
  ./tasks_and_futures : void Legion::LegionSerialization::end_task<int>(Legion::Runtime*, Legion::Internal::ContextInterface*, int*)+0x2b
  ./tasks_and_futures : void Legion::LegionTaskWrapper::legion_task_wrapper<int, &(fibonacci_task(Legion::Task const*, std::vector<Legion::PhysicalRegion, std::allocator<Legion::PhysicalRegion> > const&, Legion::Internal::TaskContext*, Legion::Runtime*))>(void const*, unsigned long, void const*, unsigned long, Realm::Processor)+0xc5
  ./tasks_and_futures : Realm::LocalTaskProcessor::execute_task(unsigned int, Realm::ByteArrayRef const&)+0x275
  ./tasks_and_futures : Realm::Task::execute_on_processor(Realm::Processor)+0x22a
  ./tasks_and_futures : Realm::UserThreadTaskScheduler::execute_task(Realm::Task*)+0x2a
  ./tasks_and_futures : Realm::ThreadedTaskScheduler::scheduler_loop()+0x2c5
  ./tasks_and_futures : void Realm::Thread::thread_entry_wrapper<Realm::ThreadedTaskScheduler, &Realm::ThreadedTaskScheduler::scheduler_loop>(void*)+0x18
  ./tasks_and_futures : Realm::UserThread::uthread_entry()+0xfc
  /lib/x86_64-linux-gnu/libc.so.6 : ()+0x49800
```
This assert only happens if the upper bits of a reservation handle are nonsensical, and it's much more rare than the original assert fail.  I was recompiling another Legion tree in the background to add some scheduling noise to the system - I'm not sure that helps or hurts.  Reassigning to Mike.

**Comment by lightsighter** [comment]
This was a different race condition.  Should be fixed with 77fb633.  Assigning back to @streichler to confirm.
